shared/js/background/newtab-tracker-stats.js

```diff
@@ -0,0 +1,348 @@
+import browser from 'webextension-polyfill'
+import constants from '../../data/constants'
+import * as browserWrapper from './wrapper.es6.js'
+import { emitter, TrackerBlockedEvent } from './before-request.es6.js'
+import tdsStorage from './storage/tds.es6'
+const { incoming, outgoing } = constants.trackerStats.events
+
+/**
+ * The extension-specific interface to tracker stats.
+ *
+ * This class handles all interaction with the extension like event handlers, alarms etc.
+ *
+ * Example:
+ *
+ * ```js
+ * const stats = new TrackerStats();
+ * const newtab = new NewTabTrackerStats(stats)
+ *
+ * await newtab.restoreFromStorage();
+ * newtab.register()
+ * ```
+ */
+export class NewTabTrackerStats {
+    /**
+     * The key to use when persisting data into storage
+     */
+    static storageKey = 'trackerStats'
+    /**
+     * The key to use when we don't want to record the company name
+     */
+    static unknownCompanyKey = 'unknown'
+    /**
+     * The prefix used for events. This is used to ensure we only handle events we care about
+     */
+    static eventPrefix = 'newTabPage_'
+
+    /**
+     * @type {Map<string, number> | null}
+     */
+    top100Companies = null
+
+    /**
+     * Internal flag to enable some certain types of logging when developing
+     */
+    _debug = false
+
+    /**
+     * @param {import("./classes/tracker-stats").TrackerStats} stats
+     */
+    constructor (stats) {
+        this.stats = stats
+    }
+
+    /**
+     * Register *all* communications with the extension here.
+     *
+     * The purpose of this method is to co-locate all extension handlers in a single
+     * place for this module.
+     */
+    register () {
+        /**
+         * This listener will redirect the request for tracker-stats.html
+         * on the new tab page to our own HTML file under `web_accessible_resources`
+         */
+        browser.webRequest.onBeforeRequest.addListener(redirectIframeForTrackerStatsMV2,
+            {
+                // todo(Shane): limit this to only the urls we care about
+                urls: ['<all_urls>']
```

Comment:
For the final version you should be able to cut down the triggering here quite a bit:
```suggestion
                urls: [`https://${constants.trackerStats.allowedOrigin}/*`],
                types: ['sub_frame']
```